### PR TITLE
Add Harmony 

### PR DESCRIPTION
**Pull request by fxfactorial**
Added Harmony to TrustWallet

### CODE REVIEW

File: include/TrustWalletCore/TWHarmonyChainID.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#pragma once
+#include "TWBase.h"
+
+TW_EXTERN_C_BEGIN
+
+/// Chain identifier for Harmony blockchain.
+TW_EXPORT_ENUM(uint32_t)
+enum TWHarmonyChainID {
+    TWHarmonyChainIDMainNet = 0x1,
+    TWHarmonyChainIDBetaNet = 0xc8,
```
Comment (vikmeup) [inline include/TrustWalletCore/TWHarmonyChainID.h]: can you remove all changes besides mainet?

Comment (vikmeup) [reply include/TrustWalletCore/TWHarmonyChainID.h]: actually we don't need these properties, networking layer would be responsible for fetching from them the node itself, can you remove them from here? feel free to hardcode them in tests only if needed

File: src/proto/Harmony.proto [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    // To Shard ID (256-bit number)
+    bytes to_shard_id = 10;
+}
+
+// Transaction signing output.
+message SigningOutput {
+    // Signed and encoded transaction bytes.
+    bytes encoded = 1;
+
+    bytes v = 2;
```
Comment (vikmeup) [inline src/proto/Harmony.proto]: @hewigovens if there a need for `v, r, s` or encoded would be enough?

File: swift/Tests/Blockchains/HarmonyTests.swift [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            $0.fromShardID = Data(hexString: "0x03")!
+            $0.toShardID = Data(hexString: "0x02")!
+            $0.payload = Data(hexString: "0x")!
+            $0.toAddress = "one1tp7xdd9ewwnmyvws96au0e7e7mz6f8hjqr3g3p"
+            $0.amount = Data(hexString: "0x0168d28e3f00280000")!
+            $0.privateKey = Data(hexString: "0xe2f88b4974ae763ca1c2db49218802c2e441293a09eaa9ab681779e05d1b7b94")!
+        }
+        let output = HarmonySigner.sign(input: input)
+        XCTAssertEqual(output.v.hexString, "0343")
+        XCTAssertEqual(output.r.hexString, "f4757c9ffad127996f788fb388be3e3e03440f6980b36dc6cee7230e390f0c13")
```
Comment (vikmeup) [inline swift/Tests/Blockchains/HarmonyTests.swift]: can you add test for encoded value?

File: tests/Harmony/SignerTests.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        /* gasLimit: */ 0x5208,
+        /* fromShardID */ 0x3,
+        /* toShardID */ 0x2,
+        /* to: */ address,
+        /* amount: */ uint256_t("0x168d28e3f00280000"),
+        /* payload: */ {});
+    auto key =
+        PrivateKey(parse_hex("0xe2f88b4974ae763ca1c2db49218802c2e441293a09eaa9ab681779e05d1b7b94"));
+    auto signer = SignerExposed(TWHarmonyChainIDLocalNet);
+    uint256_t should_be_r("0xf4757c9ffad127996f788fb388be3e3e03440f6980b36dc6cee7230e390f0c13");
```
Comment (vikmeup) [inline tests/Harmony/SignerTests.cpp]: can you add example with production transaction signing?

Comment (fxfactorial) [reply tests/Harmony/SignerTests.cpp]: Can you elaborate on what would count as production transaction signing

Comment (vikmeup) [reply tests/Harmony/SignerTests.cpp]: mainnet transaction

File: docs/coins.md [hunk 5]
```diff
@@ -48,6 +48,7 @@ This list is generated from [./coins.json](../coins.json)
 | 888     | NEO              | NEO    | <img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/neo/info/logo.png" width="32" />          | <https://neo.org>             |
 | 889     | TomoChain        | TOMO   | <img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/tomochain/info/logo.png" width="32" />    | <https://tomochain.com>       |
 | 1001    | Thunder Token    | TT     | <img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/thundertoken/info/logo.png" width="32" /> | <https://thundercore.com>     |
+| 1023    | Harmony          | ONE    | <img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/harmony/info/logo.png" width="32" />      | <https://harmony.one>         |
```
Comment (rlan35) [inline docs/coins.md]: https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/harmony/info/logo.png the image is not there.

File: src/Coin.cpp [hunk 6]
```diff
@@ -6,37 +6,38 @@
...
+#include "Semux/Address.h"
+#include "Steem/Address.h"
 #include "Stellar/Address.h"
-#include "Cosmos/Address.h"
 #include "Tezos/Address.h"
 #include "Tron/Address.h"
 #include "Wanchain/Address.h"
+#include "Waves/Address.h"
 #include "Zcash/TAddress.h"
-#include "Bravo/Address.h"
```
Comment (rlan35) [inline src/Coin.cpp]: why did we remove these?

File: coins.json [hunk 7]
```diff
@@ -1065,5 +1065,22 @@
...
+    {
+        "id": "harmony",
+        "name": "Harmony",
+        "symbol": "ONE",
+        "decimals": 18,
+        "blockchain": "Harmony",
+        "derivationPath": "m/44'/1023'/0'/0/0",
+        "curve": "secp256k1",
+        "publicKeyType": "secp256k1Extended",
+        "explorer": "https://explorer.harmony.one/#/tx/",
```
[the remark this answers is not in the document]
Comment (fxfactorial) [reply coins.json]: That's how I was shown to get a particular transaction (there is code gen based off this coins.json)

File: docs/coins.md (hunk 5 above)
Comment (fxfactorial) [reply docs/coins.md]: yes, aware: I didn't get to this repo's respective coming PR https://github.com/trustwallet/assets#repository-structure

File: src/Coin.cpp (hunk 6 above)
Comment (fxfactorial) [reply src/Coin.cpp]: Nothing removed, reordered according to clang-format, rest of code base agrees with clang-format output.

File: src/proto/Harmony.proto (hunk 2 above)
Comment (hewigovens) [reply src/proto/Harmony.proto]: I think we can combine r,s,v into signature

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/harmony/TestHarmonyTransactionSigner.kt [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+            gasPrice = ByteString.copyFrom("0x0".toHexByteArray())
+            gasLimit = ByteString.copyFrom("0x5208".toHexByteArray())
+            fromShardId = ByteString.copyFrom("0x3".toHexByteArray())
+            toShardId = ByteString.copyFrom("0x2".toHexByteArray())
+            amount = ByteString.copyFrom("0x168d28e3f00280000".toHexByteArray())
+        }
+
+        val sign: Harmony.SigningOutput = HarmonySigner.sign(signingInput.build())
+
+        assertEquals(Numeric.toHexString(sign.v.toByteArray()), "0x0343")
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/harmony/TestHarmonyTransactionSigner.kt]: can you test against `sign.encoded()`?

File: docs/coins.md (hunk 5 above)
Comment (hewigovens) [reply docs/coins.md]: can you upload your logo to  https://github.com/trustwallet/assets?

File: coins.json [hunk 9]
```diff
@@ -1065,5 +1065,22 @@
...
+        "info": {
+            "url": "https://harmony.one",
+            "client": "https://github.com/harmony-one/go-sdk",
+            "clientPublic": "https://github.com/harmony-one/go-sdk",
```
Comment (hewigovens) [inline coins.json]: do we have public node for RPC?

File: src/Harmony-One/Address.cpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
+
```
Comment (hewigovens) [inline src/Harmony-One/Address.cpp]: can we just use `Harmony` folder ? no need to add `Harmony-One`

File: src/Harmony-One/Address.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+#include "../Bech32.h"
+#include "../Hash.h"
+#include "../HexCoding.h"
+
+#include <TrezorCrypto/ecdsa.h>
+#include <TrustWalletCore/TWHRP.h>
+
+using namespace TW::Harmony;
+
+std::pair<bool, std::vector<uint8_t>> Address::isValid(const std::string &addr) {
```
Comment (hewigovens) [inline src/Harmony-One/Address.cpp]:
```suggestion
std::pair<bool, Data> Address::isValid(const std::string &addr) {
```

File: src/Harmony-One/Address.cpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+bool Address::isValid(const Data &data) {
+    return data.size() == Address::size;
+}
+
+Address::Address(const std::string &addr) {
+    const auto [success, payload] = isValid(addr);
+    if (!success) {
+        throw std::invalid_argument("address not in Harmony bech32 format");
+    }
+    Data as_base_32;
```
Comment (hewigovens) [inline src/Harmony-One/Address.cpp]: `Data converted` or use camel case

File: src/Harmony-One/Address.cpp [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+Address::Address(const Data &data) {
+    if (!isValid(data)) {
+        throw std::invalid_argument("invalid address data");
+    }
+    std::copy(data.begin(), data.end(), bytes.begin());
+}
+
+Address::Address(const PublicKey &publicKey) {
+    if (publicKey.type != TWPublicKeyTypeSECP256k1Extended) {
```
Comment (hewigovens) [inline src/Harmony-One/Address.cpp]: If compressed public key works, I think we can reuse `Cosmos::Address`, it's almost the same logic

File: src/Harmony-One/Address.cpp [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        throw std::invalid_argument("address may only be an extended SECP256k1 public key");
+    }
+    const auto data = publicKey.hash(
+        {}, static_cast<Data (*)(const byte *, const byte *)>(Hash::keccak256), true);
+    std::copy(data.end() - Address::size, data.end(), bytes.begin());
+}
+
+std::string Address::string() const {
+    Data as_base_32;
```
Comment (hewigovens) [inline src/Harmony-One/Address.cpp]: Data converted or use camel case

File: src/Harmony-One/Address.h [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+#include <vector>
+
+namespace TW::Harmony {
+
+class Address {
+  public:
+    /// Number of bytes in an address.
+    static constexpr size_t size = 20;
+
+    static constexpr auto hrp = "one";
```
Comment (hewigovens) [inline src/Harmony-One/Address.h]: Please `hrp` to `coins.json` and use `TWHRP.h`

File: src/Harmony-One/Signer.cpp [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+#include "Signer.h"
+#include "../Ethereum/RLP.h"
+#include "../HexCoding.h"
+
+using namespace TW;
+using namespace TW::Harmony;
+
+std::tuple<uint256_t, uint256_t, uint256_t> Signer::values(const uint256_t &chainID,
+                                                           const Data &signature) noexcept {
+    boost::multiprecision::uint256_t r, s, v;
```
Comment (hewigovens) [inline src/Harmony-One/Signer.cpp]: you can use methods in `uint256.h`

File: src/Harmony-One/Signer.cpp [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    return std::make_tuple(r, s, newV);
+}
+
+std::tuple<uint256_t, uint256_t, uint256_t>
+Signer::sign(const uint256_t &chainID, const PrivateKey &privateKey, const Data &hash) noexcept {
+    auto signature = privateKey.sign(hash, TWCurveSECP256k1);
+    return values(chainID, signature);
+}
+
+Proto::SigningOutput Signer::sign(const TW::Harmony::Proto::SigningInput &input) noexcept {
```
Comment (hewigovens) [inline src/Harmony-One/Signer.cpp]:
```suggestion
Proto::SigningOutput Signer::sign(const Proto::SigningInput &input) noexcept {
```

File: src/Harmony-One/Signer.cpp [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        /* payload: */ Data(input.payload().begin(), input.payload().end()));
+
+    auto signer = Signer(uint256_t(load(input.chain_id())));
+    signer.sign(key, transaction);
+    auto protoOutput = Proto::SigningOutput();
+    auto encoded = signer.rlp_no_hash(transaction, true);
+
+    protoOutput.set_encoded(encoded.data(), encoded.size());
+    auto v = store(transaction.v);
+    protoOutput.set_v(v.data(), v.size());
```
Comment (hewigovens) [inline src/Harmony-One/Signer.cpp]: let's just a signature field, we may also refactor Ethereum signing output

Comment (hewigovens) [reply src/Harmony-One/Signer.cpp]: so it's easier to handle

File: src/Harmony-One/Signer.cpp [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+void Signer::sign(const PrivateKey &privateKey, Transaction &transaction) const noexcept {
+    auto hash = this->hash(transaction);
+    auto tuple = Signer::sign(chainID, privateKey, hash);
+    transaction.r = std::get<0>(tuple);
+    transaction.s = std::get<1>(tuple);
+    transaction.v = std::get<2>(tuple);
+}
+
+Data Signer::rlp_no_hash(const Transaction &transaction, const bool include_vrs) const noexcept {
```
Comment (hewigovens) [inline src/Harmony-One/Signer.cpp]: can you use camel case? for `rlp_no_hash` and other methods

File: src/Harmony-One/Signer.cpp [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        append(encoded, RLP::encode(transaction.s));
+    } else {
+        append(encoded, RLP::encode(chainID));
+        append(encoded, RLP::encode(0));
+        append(encoded, RLP::encode(0));
+    }
+    return RLP::encodeList(encoded);
+}
+
+std::string Signer::txn_as_rlp_hex(Transaction &transaction) const noexcept {
```
Comment (hewigovens) [inline src/Harmony-One/Signer.cpp]: prefer camel case here

File: tests/interface/TWHarmonySignerTests.cpp [hunk 21]
```diff
@@ -0,0 +1,44 @@
+
```
Comment (hewigovens) [inline tests/interface/TWHarmonySignerTests.cpp]: This is just a copy of `TWEthereumSignerTests`?  Can you add a `TWHarmonySignerSign` test case? and remove duplicated `EmptyValue` and `BigInt` tests

**Comment by hewigovens** [comment]
@fxfactorial left some comments, overall is not bad

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/harmony/TestHarmonyTransactionSigner.kt (hunk 8 above)
Comment (fxfactorial) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/harmony/TestHarmonyTransactionSigner.kt]: yes, I think I have already pushed one such to Swift, will push to cover android side as well.

File: coins.json (hunk 9 above)
Comment (fxfactorial) [reply coins.json]: @rlan35 Can probably answer

File: src/Harmony-One/Address.cpp (hunk 13 above)
Comment (fxfactorial) [reply src/Harmony-One/Address.cpp]: @rlan35 can should probably provide answer if compressed public key suffices.

File: src/Harmony-One/Signer.cpp (hunk 18 above)
Comment (fxfactorial) [reply src/Harmony-One/Signer.cpp]: I am not clear if I am should be making code changes to another project's src.

File: tests/interface/TWHarmonySignerTests.cpp (hunk 21 above)
Comment (fxfactorial) [reply tests/interface/TWHarmonySignerTests.cpp]: Ah sorry, I overlooked this one.

File: src/Harmony-One/Signer.cpp (hunk 18 above)
Comment (hewigovens) [reply src/Harmony-One/Signer.cpp]: why change other project src? it seems that we don't use r,s,v inside wallet core, so we can simplify it

Comment (fxfactorial) [reply src/Harmony-One/Signer.cpp]: I thought the mentioning of "...we may also refactor Ethereum signing output" might imply me adding a commit for that work.

File: include/TrustWalletCore/TWHarmonyAddress.h [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+bool TWHarmonyAddressIsValidString(TWString *_Nonnull string);
+
+/// Creates an address from a string representaion.
+TW_EXPORT_STATIC_METHOD
+struct TWHarmonyAddress *_Nullable TWHarmonyAddressCreateWithString(TWString *_Nonnull string);
+
+/// Creates an address from a key hash.
+TW_EXPORT_STATIC_METHOD
+struct TWHarmonyAddress *_Nullable TWHarmonyAddressCreateWithKeyHash(TWData *_Nonnull keyHash);
+
```
Comment (vikmeup) [inline include/TrustWalletCore/TWHarmonyAddress.h]: @hewigovens can we reuse `CosmosAddress`? it's the same bech32 address

Comment (hewigovens) [reply include/TrustWalletCore/TWHarmonyAddress.h]: Cosmos is compressed public key

Comment (vikmeup) [reply include/TrustWalletCore/TWHarmonyAddress.h]: Oh :( then it's ok to keep as is

File: coins.json (hunk 9 above)
Comment (fxfactorial) [reply coins.json]: I checked, No public node yet for RPC

Comment (hewigovens) [reply coins.json]: we still need this when integrating into Trust Wallet, but you can provide later

**Comment by hewigovens** [comment]
created an issue for the address part: https://github.com/trustwallet/wallet-core/issues/635

**Pull request merged**